File: cli/src/main/java/com/gmail/marcosav2010/command/CommandManager.java
package com.gmail.marcosav2010.command;

import com.gmail.marcosav2010.command.base.BaseCommandRegistry;
import com.gmail.marcosav2010.logger.ILog;
import com.gmail.marcosav2010.logger.Log;
import com.gmail.marcosav2010.logger.Logger.VerboseLevel;
import com.gmail.marcosav2010.main.Main;
import com.gmail.marcosav2010.module.CommandModuleLoader;

import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.stream.Stream;

public class CommandManager {

	private final ILog log;

	private final Map<String, Command> registeredLabels;
	private final Set<Class<? extends CommandRegistry>> registries;

	public CommandManager() {
		log = new Log(Main.getInstance(), "CommandManager");
		registeredLabels = new HashMap<>();
		registries = new HashSet<>();

		register(BaseCommandRegistry.class);
		CommandModuleLoader.getInstance().flushRegistries().forEach(this::register);
	}

	public Set<Command> getCommands() {
		return Set.copyOf(registeredLabels.values());
	}

	Command fetch(String label) {
		return registeredLabels.get(label);
	}

	<T extends CommandRegistry> void register(Class<T> clazz) {
		if (registries.contains(clazz))
			return;

		registries.add(clazz);

		T r;
		try {
			r = clazz.getConstructor().newInstance();
		} catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException
				| NoSuchMethodException | SecurityException e) {
			log.log(e, "There was an error while initializing registry \"" + clazz.getName() + "\"");
			return;
		}

		int v = r.getCommands().stream().mapToInt(c -> {
			registeredLabels.put(c.getLabel(), c);
			Stream.of(c.getAliases()).forEach(a -> registeredLabels.put(a, c));
			return 1;
		}).sum();

		log.log("Loaded " + v + " commands from registry \"" + clazz.getName() + "\"", VerboseLevel.HIGH);
	}
}

File: cli/src/main/java/com/gmail/marcosav2010/command/CommandManager.kt
package com.gmail.marcosav2010.command

import com.gmail.marcosav2010.command.base.BaseCommandRegistry
import com.gmail.marcosav2010.logger.ILog
import com.gmail.marcosav2010.logger.Log
import com.gmail.marcosav2010.logger.Logger.VerboseLevel
import com.gmail.marcosav2010.main.Main
import com.gmail.marcosav2010.module.CommandModuleLoader
import java.util.*
import kotlin.reflect.KClass

class CommandManager {

    private val log: ILog
    private val registeredLabels: MutableMap<String, Command> = HashMap()
    private val registries: MutableSet<KClass<out CommandRegistry>>
    val commands: Set<Command>
        get() = registeredLabels.values.toSet()

    init {
        log = Log(Main.instance, "CommandManager")

        registries = HashSet()
        register(BaseCommandRegistry::class)
        CommandModuleLoader.flushRegistries().forEach { clazz -> register(clazz) }
    }

    fun fetch(label: String) = registeredLabels[label] ?: throw CommandNotFoundException("Command not found")

    private fun <T : CommandRegistry> register(clazz: KClass<T>) {
        if (registries.contains(clazz)) return

        registries.add(clazz)

        val r: CommandRegistry
        try {
            r = clazz.constructors.first().call()
        } catch (e: Exception) {
            log.log(e, "There was an error while initializing registry \"" + clazz.qualifiedName + "\"")
            return
        }

        val v = r.commands.map { c ->
            registeredLabels[c.label] = c
            c.aliases.forEach { a -> registeredLabels[a] = c }
            1
        }.sum()

        log.log("Loaded " + v + " commands from registry \"" + clazz.simpleName + "\"", VerboseLevel.HIGH)
    }
}